internal/api/cluster_installation.go

```diff
@@ -359,3 +364,240 @@ func handleRunClusterInstallationMattermostCLI(c *Context, w http.ResponseWriter
+		return
+	}
+	if clusterInstallations == nil {
+		c.Logger.WithError(err).Error("no matching cluster installations found")
+		w.WriteHeader(http.StatusNotFound)
+		return
+	}
+
+	// verify that the allows installation is false for the source cluster before migration starts
+	cluster, err := c.Store.GetCluster(mcir.ClusterID)
+	if err != nil {
+		c.Logger.WithError(err).Error("failed to get cluster")
+		w.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+
+	// Migrate the cluster installations to the target cluster
+	if !cluster.AllowInstallations {
+		c.Logger.Infof("migrating installation(s) to clusterID: %s", mcir.TargetCluster)
+		err = c.Store.MigrateClusterInstallations(clusterInstallations, mcir.TargetCluster)
+		if err != nil {
+			c.Logger.WithError(err).Error("failed to migrate cluster installation(s)")
+			w.WriteHeader(http.StatusInternalServerError)
+			return
+		}
+	} else {
+		c.Logger.WithError(err).Error("allow installation must be set to false for the source cluster.")
+		w.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+
+	w.WriteHeader(http.StatusOK)
+}
+
+// handleMigrateDns responds to Post /api/cluster_installation/migrate/dns.
+func handleMigrateDNS(c *Context, w http.ResponseWriter, r *http.Request) {
+	mcir, err := model.NewMigrateClusterInstallationRequestFromReader(r.Body)
+	if err != nil {
+		c.Logger.WithError(err).Error("failed to decode cluster migration request")
+		w.WriteHeader(http.StatusBadRequest)
+		return
+	}
+	c.Logger = c.Logger.WithField("cluster_installation", mcir)
+
+	// Reset the DNS configuration status for respective installations to update the CNAME with the new LB.
+	filter := &model.ClusterInstallationFilter{
+		ClusterID:      mcir.ClusterID,
+		InstallationID: mcir.InstallationID,
+		Paging:         model.AllPagesNotDeleted(),
+	}
+	clusterInstallations, err := c.Store.GetClusterInstallations(filter)
+	if err != nil {
+		c.Logger.WithError(err).Error("failed to query cluster installations")
+		w.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+	if clusterInstallations == nil {
+		c.Logger.WithError(err).Error("no matching cluster installations found for ")
+		w.WriteHeader(http.StatusNotFound)
+		return
+	}
+	filter = &model.ClusterInstallationFilter{
+		ClusterID:      mcir.TargetCluster,
+		InstallationID: mcir.InstallationID,
+		Paging:         model.AllPagesNotDeleted(),
+	}
+	newClusterInstallations, err := c.Store.GetClusterInstallations(filter)
+	if err != nil {
+		c.Logger.WithError(err).Error("failed to query cluster installations")
+		w.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+	if newClusterInstallations == nil {
+		c.Logger.WithError(err).Error("no matching cluster installations found for")
+		w.WriteHeader(http.StatusNotFound)
+		return
+	}
+
+	// verify that the allows installation is false for the source cluster before migration starts
+	cluster, err := c.Store.GetCluster(mcir.ClusterID)
+	if err != nil {
```

Comment:
Unfortunately `cluster` here still can be `nil` even if no error occurred. This happens when such cluster does not exist at all.

We need to handle this case as otherwise the line bellow that references `cluster.AllowInstallation` causes a panic.

Reply:
Yes, Good point. fixed